### ISSUE TITLE
ST11 - False positive with array of nested fields in structs in BigQuery

### ISSUE THREAD

**Issue by chris-pomelo** [sqlfluff/sqlfluff#6997]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

There is a false ST11 error when using nested fields with `unnest` on an array of nested structs.

Similar to https://github.com/sqlfluff/sqlfluff/issues/6558 but with a deeper level of nesting. `level1.level2.level3` where `level1` is the unnested array.

### Expected Behaviour

No error in the example below.

### Observed Behaviour

Error message:
```
L:  13 | P:  11 | ST11 | Joined table 'unnest(test_table.generic_array) as g' not
                       | referenced elsewhere in query
                       | [structure.unused_join]
```

### How to reproduce

To reproduce, simply run a SQLFluff lint command on the following SQL.

```sql
with
    test_table as (
        select
            1 as id,
            array<struct<repeated_structs struct<nested_struct struct<generic_field string>>>>[
                struct(struct(struct('abc' as generic_field))),
                struct(struct(struct('def' as generic_field)))
            ] as generic_array
    )

select test_table.id, g.repeated_structs.nested_struct.generic_field
from test_table
left join unnest(test_table.generic_array) as g
```

### Dialect

BigQuery

### Version

sqlfluff, version 3.4.1
Python 3.11.10

### Configuration

```
[sqlfluff]

dialect = bigquery

templater = jinja

exclude_rules = ambiguous.column_count, structure.column_order,
  layout.indent, layout.cte_bracket, layout.select_targets, layout.spacing,
  layout.long_lines,
  layout.cte_newline,
  ambiguous.join,
  convention.select_trailing_comma,
  aliasing.length,
  layout.end_of_file,
  layout.keyword_newline

max_line_length = 0

processes = -1
large_file_skip_byte_limit = 0

[sqlfluff:templater:jinja]
apply_dbt_builtins = True
load_macros_from_path = data/dbt/pomelo/macros,data/utils/sqlfluff/sqlfluff_macros
library_path = data/utils/sqlfluff/sqlfluff_libs

[sqlfluff:templater:dbt]
project_dir = data/pomelo/dbt

[sqlfluff:indentation]
allow_implicit_indents = True
indented_ctes = True

[sqlfluff:rules:aliasing.length]
min_alias_length = 3

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = lower
[sqlfluff:rules:capitalisation.identifiers]
extended_capitalisation_policy = lower
[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = lower
[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower
[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = lower

[sqlfluff:rules:convention.not_equal]
preferred_not_equal_style = c_style

[sqlfluff:rules:convention.terminator]
multiline_newline = True

[sqlfluff:rules:convention.quoted_literals]
preferred_quoted_literal_style = single_quotes

[sqlfluff:rules:references.special_chars]
allow_space_in_identifier = True

[sqlfluff:rules:references.qualification]
subqueries_ignore_external_references = True
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by adhavan18** [comment]
This does not reproduce on `main` any more. I ran the exact SQL from the report against 4.2.2:

```text
$ sqlfluff lint --dialect bigquery --rules ST11 issue_6997.sql
All Finished!
```

I checked the rule was actually running rather than silently skipping, and it still fires where it should:

```text
$ sqlfluff lint --dialect bigquery --rules ST11 unused_join.sql
L:   3 | P:  11 | ST11 | Joined table 'tbl_b as b' not referenced elsewhere in query
```

Also clean at two and four levels of nesting, so it is not depth-specific.

The fix looks to be #7795 (struct/nested-field reference handling), merged in April. #7007 was the earlier attempt here and was closed while still WIP, so this issue just outlived its fix.

`ST11.yml` covers the two halves separately (`test_pass_table_expression_function_6558` for the UNNEST chain, `test_pass_deeply_nested_struct_field_access` for deep struct access on a plain join) but not the combination reported here, so I opened #8240 to pin exactly this shape. It fails on the commit before #7795 and passes on `main`.

Safe to close once that lands, or now if you would rather not wait on it.
